### PR TITLE
alert component and service

### PR DESCRIPTION
**Pull request by tomeraronovsky**
alert component and service
animations & new close button next prs

### CODE REVIEW

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss [hunk 1]
```diff
@@ -0,0 +1,39 @@
+@import '../../style/colors';
+@import '../../style/typography';
+
+::ng-deep {
```
Comment (ishaiborovoy) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: I think better to use host too, see other components.  So the case will be isolated

Comment (omrihecht) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: @tomeraronovsky , ::ng-deep in this way is exposed to all always, this shouldn't be anywhere that is not supposed to shared everywhere.
good catch @ishaiborovoy

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  .b-alert {
+    background-color: $white;
+    border-radius: 4px;
+    box-shadow: 0 5px 10px 0 rgba(83,83,83,0.2);
+    min-width: 395px;
+    min-height: 78px;
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: i don't think you need min-height.
height should be determined by content
also width should be set as a const (or enum),  min-width could also mean the content can be 100vw, i don't think it's what they meant.
please use mixin for box-shadow (look for it in the code), i think there is also one for border-radius

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<ng-template #alertTemplateRef>
+  <div class="alert-container">
+    <b-icon class="close-icon" [icon]="closeIcon" [color]="closeColor" [size]="closeSize"></b-icon>
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.html]: use button where the element has button behaviour

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    .alert-container {
+      position: relative;
+      display: flex;
+      flex-direction: row;
+      align-items: center;
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: no need to center align vertically, add padding to top and bottom instead

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import {Component, TemplateRef, ViewChild, ViewContainerRef} from '@angular/core';
+import {cloneDeep, Dictionary} from 'lodash';
+import {IconColor, Icons, IconSize} from '../../icons/icons.enum';
+import {InfoStripIcon} from '../info-strip/info-strip.types';
+import {AlertConfig} from './alert.interface';
+import {Overlay, OverlayConfig, OverlayRef} from '@angular/cdk/overlay';
+import {TemplatePortal} from '@angular/cdk/portal';
+
+const ALERT_DURATION = 3000;
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.ts]: i think it's 7000, check hibob-web popup service (`showPositiveMessage`)

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.enum.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export enum AlertType {
+  information = 'information',
+  warning = 'warning',
+  error = 'error',
+  success = 'success'
+}
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.enum.ts]: 👍

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.module.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import {AlertService} from './alert-service/alert.service';
+import {BrowserAnimationsModule, NoopAnimationsModule} from '@angular/platform-browser/animations';
+
+@NgModule({
+  declarations: [AlertComponent],
+  entryComponents: [AlertComponent],
+  imports: [
+    CommonModule,
+    IconsModule,
+    NoopAnimationsModule,
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.module.ts]: why is this here?
`NoopAnimationsModule`

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss (hunk 1 above)
Comment (tomeraronovsky) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: I wrapped the b-alert with ng-deep because b-alert is the panelClass, and I saw that every panelClass we wrap with ng-deep

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.html (hunk 3 above)
Comment (tomeraronovsky) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.html]: as we talked I need to change some settings for the X button, i will add it next pr, for now i gonna remove it

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss (hunk 4 above)
Comment (tomeraronovsky) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: didn't understand why, if you align it vertically it ensure it will looks good in the container

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.module.ts (hunk 7 above)
Comment (tomeraronovsky) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.module.ts]: wrong, remove it

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  private overlayConfig: OverlayConfig;
+  private templatePortal: TemplatePortal;
+  public overlayRef: OverlayRef;
+
+  constructor(private componentFactoryResolver: ComponentFactoryResolver,
+              private injector: Injector,
+              private overlay: Overlay) { }
+
+  public showAlert(config: AlertConfig): void {
+    this.injectAlertComponent();
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.ts]: 👍

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  private injectAlertComponent(): void {
+    const componentFactory = this.componentFactoryResolver.resolveComponentFactory(AlertComponent);
+    this.alertComponentRef = componentFactory.create(this.injector);
+  }
+
+  private getConfig(): OverlayConfig {
+    const positionStrategy = this.overlay.position().global()
+      .centerHorizontally().top('20px');
+    const panelClass = 'b-alert';
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.ts]: rename `b-alert-panel`, the class will be added to the material generated panel, not the b-alert component

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    const backdropClass = '';
+    return {
+      disposeOnNavigation: true,
+      hasBackdrop: false,
+      backdropClass,
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.ts]: if you don't have backdrop, no need for backdrop class

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.html [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<ng-template #alertTemplateRef>
+  <div class="alert-container">
+    <b-icon class="icon" [icon]="iconsDic[alertConfig.alertType].icon" [color]="iconsDic[alertConfig.alertType].color" [size]="iconSize"></b-icon>
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.html]: why do you need ng-template?
also you can remove alert-container, and use the host as the container

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss [hunk 12]
```diff
@@ -0,0 +1,42 @@
+@import '../../style/colors';
+@import '../../style/typography';
+@import '../../style/variables';
+
+$alert_width: 395px;
+
+::ng-deep {
+  .b-alert {
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: b-alert-panel

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss (hunk 2 above)
Comment (omrihecht) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: please remove height,
use shadow mixin,
remove min-width (should be width, f i remember correctly, you can add to the panel config)
if you want let's have a short session and do the css right

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss (hunk 4 above)
Comment (omrihecht) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.scss]: because the container should be determined by the content height, not the other way around

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+@Component({
+  selector: 'b-alert',
+  templateUrl: './alert.component.html',
+  styleUrls: ['./alert.component.scss'],
+})
+export class AlertComponent {
+  @ViewChild('alertTemplateRef') alertTemplateRef: TemplateRef<any>;
+  public alertConfig: AlertConfig;
+  public closeAlertCallback: Function;
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.ts]: doesn't look like should be public

Comment (tomeraronovsky) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.ts]: it must be public for binding it in the service

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.spec.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      const ALERT_DURATION_TICK = 7001;
+      const alertConfig: AlertConfig = {
+        alertType: AlertType.success,
+        text: 'text',
+        title: 'title'
+      };
+      alertService.showAlert(alertConfig);
+      tick(ALERT_DURATION_TICK);
+      expect(alertService.overlayRef.dispose).toHaveBeenCalled();
+    }));
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.spec.ts]: great!

Comment (omrihecht) [reply projects/ui-framework/src/lib/buttons-indicators/alert/alert-service/alert.service.spec.ts]: it's good practice to have tests for every public method, you can add a small one to `closeAlert`

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+      alertType: AlertType.success,
+      text: 'text',
+      title: 'title'
+    };
+    fixture.detectChanges();
+  });
+
+  it('should call to close alert callback', () => {
+    component.closeAlertCallback = jasmine.createSpy();
+    component.closeAlert();
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts]: a better test would be to click the close button, not invoke the method

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+      providers: [
+        { provide: AlertService, useValue: {} },
+      ]
+    })
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(AlertComponent);
+    component = fixture.componentInstance;
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts]: add `spyOn(component, 'closeAlertCallback');`

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    component.alertConfig = {
+      alertType: AlertType.success,
+      text: 'text',
+      title: 'title'
+    };
+    fixture.detectChanges();
+  });
+
+  it('should call to close alert callback', () => {
+    component.closeAlertCallback = jasmine.createSpy();
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts]: remove

File: projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    component.closeAlert();
+    expect(component.closeAlertCallback).toHaveBeenCalled();
+  });
+});
```
Comment (omrihecht) [inline projects/ui-framework/src/lib/buttons-indicators/alert/alert.component.spec.ts]: you should also test the config to dom implementation (the integration test)

File: projects/ui-framework/src/public_api.ts [hunk 19]
```diff
@@ -81,6 +81,10 @@ export { LinkColor, LinkTarget } from './lib/buttons-indicators/link/link.enum';
 // InfoTooltip Module
 export { InfoTooltipModule } from './lib/popups/info-tooltip/info-tooltip.module';
 export { InfoTooltipComponent } from './lib/popups/info-tooltip/info-tooltip.component';
+// Alert Module
+export { AlertService } from './lib/buttons-indicators/alert/alert-service/alert.service';
+export { AlertModule } from './lib/buttons-indicators/alert/alert.module';
+export { AlertConfig } from './lib/buttons-indicators/alert/alert.interface';
```
Comment (omrihecht) [inline projects/ui-framework/src/public_api.ts]: you also need to export the AlertComponent, even though it's part of the module

**Comment by omrihecht** [comment]
you need to update/rebase to merge
